### PR TITLE
GG-24505 [IGNITE-12099] Don't write metadata to disk in discovery thread

### PR DESCRIPTION
**Pull request by ibessonov**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java [hunk 1]
```diff
@@ -54,6 +63,12 @@
     /** */
     private final IgniteLogger log;
 
+    /** */
+    private BinaryMetadataAsyncWriter writer;
+
+    /** */
+    private final ConcurrentHashMap<WriteOpSync, GridFutureAdapter> writeOpFutures = new ConcurrentHashMap<>();
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java]: Type of the field can be changed to ConcurrentMap

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java [hunk 2]
```diff
@@ -89,6 +104,17 @@
...
+
+        writer = new BinaryMetadataAsyncWriter();
+        new IgniteThread(writer).start();
+    }
+
+    /**
+     * Stops worker for async writing of binary metadata.
+     */
+    void stop() {
+        if (writer != null)
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java]: Looks like "writer" cannot be null.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java [hunk 3]
```diff
@@ -117,6 +143,10 @@ void writeMetadata(BinaryMetadata binMeta) {
 
             U.error(log, msg);
 
+            for (GridFutureAdapter fut : writeOpFutures.values()) {
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java]: Please remove braces.

Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java]: How is it in sync with writeMetadataAsync? Can we add new future while iterating?
Also, shouldn't we clear this map after? Does it matter?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java [hunk 4]
```diff
@@ -182,4 +212,160 @@ private BinaryMetadata readMetadata(int typeId) {
...
+
+        /** */
+        private WriteOpSync(int typeId, int typeVer) {
+            this.typeId = typeId;
+            this.typeVer = typeVer;
+        }
+
+        /** {@inheritDoc} */
+        @Override public int hashCode() {
+            return Objects.hash(typeId, typeVer);
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java]: Too allocation-heavy for a hashCode, please use something like "31 * typeId + typeVer" or even "typeId ^ typeVer". Up to you.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java [hunk 5]
```diff
@@ -182,4 +212,160 @@ private BinaryMetadata readMetadata(int typeId) {
...
+        }
+
+        /** {@inheritDoc} */
+        @Override public boolean equals(Object obj) {
+            if (obj == this)
```
Comment (ibessonov) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java]: Looking at your code I guarantee that this condition is always false.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsBinaryMetadataAsyncWritingTest.java [hunk 6]
```diff
@@ -0,0 +1,729 @@
...
+
+        IgniteCache<Object, Object> cache = ig0.cache(DEFAULT_CACHE_NAME);
+
+        cache.put(ig1Key, new TestAddress(0, "USA", "NYC", "6th Ave"));
+
+        waitForTopology(1, 10_000);
+    }
+
+    /**
+     * Verifies that
```
Comment (ibessonov) [inline modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsBinaryMetadataAsyncWritingTest.java]: Not completed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java (hunk 1 above)
Comment (sergey-chugunov-1985) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java (hunk 2 above)
Comment (sergey-chugunov-1985) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java]: Removed.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java (hunk 3 above)
Comment (sergey-chugunov-1985) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java]: Done.

Comment (sergey-chugunov-1985) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java]: Indeed here was a data race, I closed it with adding additional flag similar to implementation in BinaryMetadataTransport.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java (hunk 4 above)
Comment (sergey-chugunov-1985) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java]: Done.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java (hunk 5 above)
Comment (sergey-chugunov-1985) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/binary/BinaryMetadataFileStore.java]: Removed.

File: modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsBinaryMetadataAsyncWritingTest.java (hunk 6 above)
Comment (sergey-chugunov-1985) [reply modules/core/src/test/java/org/apache/ignite/internal/processors/cache/persistence/IgnitePdsBinaryMetadataAsyncWritingTest.java]: Finished.

**Pull request closed without merging**